syntax = "proto3";

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";

enum ExampleEnumModel {
    ENUM0 = 0;
    ENUM1 = 1;
    ENUM2 = 2;
}

message ExampleNestedModel {
    string string_key = 1;
    int32 int32_key = 2;
}

message ExampleDBModel {
    // Simple type
    int32 int32_key = 1;
    string string_key = 2;
    bool bool_key = 3;
    bytes bytes_key = 4;

    // Container types with simple values
    map<string, string> map_string_string = 5;
    map<string, int32> map_string_int32 = 6;
    repeated string string_array_key = 7;
    repeated int32 int32_array_key = 8;

    // Container types with complex values
    repeated ExampleNestedModel complex_array_key = 9;

    // Enum types
    ExampleEnumModel enum_key = 10;

    // Complex types from protobuf stdlib
    google.protobuf.Timestamp timestamp_key = 11;
    google.protobuf.Struct struct_key = 12;

    // Other special types
    google.protobuf.NullValue null_key = 13;
}
